### PR TITLE
Allow streaming diffs in constant memory

### PR DESCRIPTION
**Pull request by ChrisPenner**
## Overview

The current v2 diff implementation is more efficient than the old one, but still loads the whole diff into memory. This causes a lot of memory bloat on Share where we load a full namespace diff when updating name lookup indexes from scratch. E.g. I tried rebuilding the index for stew's  public share codebase and I hit 60GB of memory locally before my mac killed the process 😅.

There's really no need to load any of these into memory since we can stream the diff directly into SQLite to build the index that way, which uses constant memory instead. This greatly improve's Share's memory usage profile and also allows those of us running on Macbook Air's to build large indexes locally again.

On this new branch I was able to re-compute  the index for Stew's codebase without memory increasing more than 10  MB past the  282MB baseline 🎉 . Unfortunately re-computing these indexes from scratch is still computationally intensive (it took 12 minutes on my macbook air 😢 ) but that's a different problem to solve. In most cases we  don't need to compute them from scratch (unless it's a new branch, where it will hopefully be a small index); Stew's  codebase is hopefully the worst-case we'll ever see, since projects will mean each codebase is _significantly_ smaller since it won't contain multiple projects and multiple releases for each of their projects AND copies of all their dependencies 😓 

## Implementation notes

* Switches `TreeDiff` to be monadic and only compute one layer at a time.
* Adds `streamNameChanges` consumes the TreeDiff one layer at a time and streams the diff to a callback as it goes.
* Changes the name lookup builder to use the streaming method.

For now I left the command for namespace patch to 

## Interesting/controversial decisions

As always, there are many different ways to implement streaming in Haskell. I went with a simple callback passing approach for this one because most channel or STM approaches require IO and forking (which  requires MonadUnliftIO); which means we'd need to interleave a bunch of IO into our transaction and provide things like `Transaction.unsafeUnliftIO`, then fork a worker thread, then co-ordinate with that worker so it's shut-down correctly and the sqlite writer knows when to stop listening for changes. It's all possible of course, but is more complex, more annoying, more error-prone and less safe.

This version lets us just stay in Transaction and easily handle one chunk of diffs at a time.

## Test coverage

Tested rebuilding name lookups locally; 

Codebase containing `base` and `distributed`:

* Trunk: 18s  to rebuild, memory spiked from 282 MB -> 1720 MB
* Streaming version: 20s to rebuild, memory stayed at the baseline of 282 MB

So it's a bit slower, but uses significantly less memory. I think this is worth the tradeoff. Lower memory residency may mean quicker GC pauses too because there are fewer live objects to copy over.

For Stew's monolithic codebase:

* Trunk: DNF; after ~30min memory had spiked to 60GB and my mac gave up and  killed the process.
* Streaming version: ~12 mins, memory stayed flat at the baseline of 282 MB

### CODE REVIEW

File: parser-typechecker/src/Unison/Codebase/SqliteCodebase/Operations.hs [hunk 1]
```diff
@@ -643,12 +643,18 @@ ensureNameLookupForBranchHash getDeclType mayFromBranchHash toBranchHash = do
...
-      termNameAddsWithCT <- do
-        for termNameAdds \(name, ref) -> do
-          refWithCT <- addReferentCT ref
-          pure $ toNamedRef (name, refWithCT)
-      Ops.buildNameLookupForBranchHash mayExistingLookupBH toBranchHash (termNameAddsWithCT, toNamedRef <$> termNameRemovals) (toNamedRef <$> typeNameAdds, toNamedRef <$> typeNameRemovals)
+      -- TODO: Might be worth switching to a streaming lib to make this less confusing.
+      Ops.buildNameLookupForBranchHash
+        mayExistingLookupBH
+        toBranchHash
+        ( \save -> do
```
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/Codebase/SqliteCodebase/Operations.hs:650]: The nested callback stuff is a bit annoying;

Totally open to other options, e.g. using a TBQueue or something, but note that those are also slightly annoying because we'd then want to fork a separate worker, and doing unlifting and IO is annoying in its own right since we don't want to add a MonadUnliftIO instance for Transaction.

TLDR; this is a little mindbending but seemed like the safest and lesser of the evils, it allows us to stay in Transaction the whole time and means we'll keep at most one namespace diff in memory at a time.

**Review by pchiusano — APPROVED** [review]
Nice!

File: parser-typechecker/src/U/Codebase/Branch/Diff.hs [hunk 2]
```diff
@@ -113,34 +117,37 @@ instance Semigroup NameBasedDiff where
...
+              | otherwise -> do
+                  fromChildBranch <- Causal.value fromC
+                  toChildBranch <- Causal.value toC
+                  diffBranches fromChildBranch toChildBranch >>= \case
+                    TreeDiff (defDiffs :< Compose mchildren) -> do
+                      children <- mchildren
+                      if (isEmptyDefDiffs defDiffs && null children)
+                        then pure Nothing
+                        else pure . Just $ (defDiffs :< Compose (pure children))
+  pure $ TreeDiff (defDiff :< Compose childDiff)
```
Comment (tstat) [inline parser-typechecker/src/U/Codebase/Branch/Diff.hs]: It looks like `diffBranches` should return a `TreeDiff m` instead of `m (TreeDiff m)`.

Comment (ChrisPenner) [reply parser-typechecker/src/U/Codebase/Branch/Diff.hs]: Great catch 👍🏼

File: parser-typechecker/src/U/Codebase/Branch/Diff.hs [hunk 3]
```diff
@@ -113,34 +117,37 @@ instance Semigroup NameBasedDiff where
...
+            These fromC toC
+              | Causal.valueHash fromC == Causal.valueHash toC -> do
+                  -- This child didn't change.
+                  pure Nothing
+              | otherwise -> do
+                  fromChildBranch <- Causal.value fromC
+                  toChildBranch <- Causal.value toC
+                  diffBranches fromChildBranch toChildBranch >>= \case
+                    TreeDiff (defDiffs :< Compose mchildren) -> do
+                      children <- mchildren
```
Comment (tstat) [inline parser-typechecker/src/U/Codebase/Branch/Diff.hs]: I'm surprised to see the `mchildren` effects being performed here. Forcing a child forces its children? Am I reading that correctly? I feel like I must be missing something but that reads to me like it would force of the whole tree whenever the first child is forced.

Comment (mitchellwrosen) [reply parser-typechecker/src/U/Codebase/Branch/Diff.hs]: I agree with this comment :thinking:

Comment (ChrisPenner) [reply parser-typechecker/src/U/Codebase/Branch/Diff.hs]: Strange, in practice it seems to work just fine, but I agree this is wrong. I think it might just not be triggering this case often enough to affect the memory footprint, but at any rate I'll switch it around and see if it works out 👍🏼 

This was meant to be an optimization so that you don't see child trees that don't have a diff in them, but I think as it is it's just a bug, I'll swap that around 😄

Comment (mitchellwrosen) [reply parser-typechecker/src/U/Codebase/Branch/Diff.hs]: Ok. And is it still clear to you, despite this bug, why we see memory improvements? I had thought that calculating a diff one layer at a time was the source of the streaming.

Comment (ChrisPenner) [reply parser-typechecker/src/U/Codebase/Branch/Diff.hs]: Okay I dug into this a bit and figured it out. The use-case which I cared the most about was building a name lookup from scratch, since that's the use-case that uses the most memory. But in that situation we're diffing against the empty branch, so we never hit this case where we have to recursively diff children at the same position. Meaning we get all the one-layer-at-a-time streaming benefits in this case, which is why my tests show such an improvement despite this bug.

The problematic case would only be triggered on subsequent diffs, and usually in those situations the diff is small enough that memory isn't as much of a concern; but if someone pushed a very large diff that shared a spine with the previous branch then it would trigger this and use more memory than needed.

So TLDR; this problem would've shown up as a bit of extra memory usage in subsequent pushes where the from and to branches were close to each other, I patched it up 😄 

Great catch on this one @tstat!

**Review by mitchellwrosen — APPROVED** [review]

File: parser-typechecker/src/U/Codebase/Branch/Diff.hs [hunk 4]
```diff
@@ -113,34 +114,34 @@ instance Semigroup NameBasedDiff where
...
+              -- prefix query, this would be faster than crawling to get all the deletes.
+              removedChildBranch <- Causal.value ca
+              pure . Just . unTreeDiff $ diffBranches removedChildBranch Branch.empty
+            That ca -> do
+              newChildBranch <- Causal.value ca
+              pure . Just . unTreeDiff $ diffBranches Branch.empty newChildBranch
+            These fromC toC
+              | Causal.valueHash fromC == Causal.valueHash toC -> do
+                  -- This child didn't change.
+                  pure Nothing
```
Comment (tstat) [inline parser-typechecker/src/U/Codebase/Branch/Diff.hs]: Doesn't need to hold up the pr, but since determining if the key should be included doesn't require doing anything in `m`, couldn't we use `mapMaybe` rather than `wither`? Then `TreeDiff` could be `Cofree (Compose (Map NameSegment) m)` rather than `Cofree (Compose m (Map NameSegment))`. This would allow us to  load one namespace at a time rather than one level at a time. So, we would only run up to two `Causal.value`s at a time rather than running the sum of the number of immediate namespaces in both maps.

Comment (ChrisPenner) [reply parser-typechecker/src/U/Codebase/Branch/Diff.hs]: Yeah this is what I was attempting to do with my original flawed approach 😓 , but forgot to add it back now that things are moved around.

Just implemented your suggestion and seems good 😄
